effekt/shared/src/main/scala/effekt/source/Tree.scala

```diff
@@ -525,68 +529,73 @@ export MatchPattern.*
- *     │  │─ [[ ValueTypeRef ]]
- *     │
- *     │─ [[ BlockType ]]
- *     │  │─ [[ BlockTypeTree ]]
- *     │  │─ [[ FunctionType ]]
- *     │  │─ [[ BlockTypeRef ]]
- *     │
+ *     │─ [[ ValueTypeTree ]]
+ *     │─ [[ BlockTypeTree ]]
+ *     │─ [[ TypeRef ]]
+ *     │─ [[ BoxedType ]]
+ *     │─ [[ FunctionType ]]
+ *     │─ [[ Effectful ]]
  *
  * --------------------------------------------
  */
 sealed trait Type extends Tree
 
 /**
- * Value Types
+ * Trees that represent inferred or synthesized *value* types (not present in the source = not parsed)
  */
-enum ValueType extends Type {
+case class ValueTypeTree(tpe: symbols.ValueType) extends Type
 
-  /**
-   * Trees that represent inferred or synthesized types (not present in the source)
-   */
-  case ValueTypeTree(tpe: symbols.ValueType)
+/**
+ * Trees that represent inferred or synthesized *block* types (not present in the source = not parsed)
+ */
+case class BlockTypeTree(eff: symbols.BlockType) extends Type
 
-  /**
-   * Types of first-class functions
-   */
-  case BoxedType(tpe: BlockType, capt: CaptureSet)
+/*
+ * Reference to a type, potentially with bound occurences in `args`
+ */
+case class TypeRef(id: IdRef, args: List[ValueType]) extends Type, Reference
 
-  // Bound occurrences (args can be empty)
-  case ValueTypeRef(id: IdRef, args: List[ValueType]) extends ValueType, Reference
-}
-export ValueType.*
+/**
+ * Types of first-class computations
+ */
+case class BoxedType(tpe: BlockType, capt: CaptureSet) extends Type
 
 /**
- * Block Types
+ * Types of (second-class) functions
  */
-enum BlockType extends Type {
+case class FunctionType(tparams: List[Id], vparams: List[ValueType], bparams: List[(Option[IdDef], BlockType)], result: ValueType, effects: Effects) extends Type
 
-  /**
-   * Trees that represent inferred or synthesized types (not present in the source)
-   */
-  case BlockTypeTree(eff: symbols.BlockType)
-  case FunctionType(tparams: List[Id], vparams: List[ValueType], bparams: List[(Option[IdDef], BlockType)], result: ValueType, effects: Effects)
-  case BlockTypeRef(id: IdRef, args: List[ValueType]) extends BlockType, Reference
+/**
+ * Type-and-effect annotations
+ */
+case class Effectful(tpe: ValueType, eff: Effects) extends Type
+
+// These are just type aliases for documentation purposes.
+type BlockType = Type
+type ValueType = Type
+
+// TODO(jiribenes, 2024-04-22): Where should this go?
+//     We only use this in Namer, should it be there as a private def `describeType`?
+def describe(t: Type)(using C: Context): String = t match {
+  case _: TypeRef => s"a type reference ${t.sourceOf}"
+  case _: BoxedType => s"a boxed type ${t.sourceOf}"
+  case _: FunctionType => s"a second-class function type ${t.sourceOf}"
+  case _: Effectful => s"a type-and-effect annotation ${t.sourceOf}"
+
+  // THESE TWO SHOULD NEVER BE USER-VISIBLE!
+  case ValueTypeTree(tpe) => s"a value type tree ${tpe}"
+  case BlockTypeTree(eff) => s"a block type tree ${eff}"
 }
```

Comment:
Where does this go? I think that a `private def` in `Namer` would make the most sense?

Reply:
Moved to Namer.